Add stdenv & CC support for Zig

RossComputerGuy opened the pull request:
## Description of changes

Introduces the ability to compile programs in nixpkgs using Zig with its own `zigStdenv` and CC adapter. This is kinda broken at the moment but it does not break anything critical so rebuilds should be 0 (except patching things like meson). However, the only problem seems to be an issue with Zig trying to link glibc right now. It is linking the `libc.so` linker script instead of the shared library.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RossComputerGuy committed:
zig: add cc wrapper

RossComputerGuy committed:
cc-wrapper: add zig

RossComputerGuy committed:
pkgs/stdenv/cross: add zig

RossComputerGuy committed:
pkgs/top-level/{release,stage}.nix: add zig package set

RossComputerGuy committed:
zig: add stdenv

RossComputerGuy committed:
meson: include patch for zig cc

RossComputerGuy commented:
Close to this working, need to resolve a crash discovered that is being caused by the CC wrapper.

RossComputerGuy commented:
We're now producing binaries correctly. `pkgsZig.hello` builds.

RossComputerGuy commented:
Fixed eval and so this should work now properly.

RossComputerGuy committed:
sqlite: fix build issue with zig cc

Ericson2314 commented:
I'm not against this, but what exactly is the point? Isn't Zig just wrapping clang? So now we have double-wrapped clang?

RossComputerGuy commented:
Zig has its own compiler backend plus I'm sure that Zig offers a bit more optimizations and size saving measures. One of my ideas is we could expand this and use Zig prebuilds to bootstrap nixpkgs since it's a static binary.

Ericson2314 commented:
Does Zig have it's own compiler backend for *C and C++* though? I would be very surprised, to say the least, if they managed to rig up clang to a different backend or write their own C/C++ compiler,

> One of my ideas is we could expand this and use Zig prebuilds to bootstrap nixpkgs since it's a static binary.

I do want to replace our current bootstrap tools set with just static binaries. I would hope after the FreeBSD native bootstrapped is upstreamed to maybe take a stab at at with Clang.

RossComputerGuy commented:
> Does Zig have it's own compiler backend for _C and C++_ though? I would be very surprised, to say the least, if they managed to rig up clang to a different backend or write their own C/C++ compiler,

Idk about C++ but there is Aro for C and I'm pretty sure that's been added into Zig itself.

> I do want to replace our current bootstrap tools set with just static binaries. I would hope after the FreeBSD native bootstrapped is upstreamed to maybe take a stab at at with Clang.

Sounds interesting, my idea with Zig bootstrapping is to use the static build of Zig. From there we could put together a minimal host tools environment inside of Nix without having to have what we currently have. If we're able to build just a minimal set of host tools which then could be used to build other things more complete then we could bootstrap more inside of the Nix store.

Ericson2314 commented:
Could we start with CC wrapper for Aro then? I rather first gain experience with the underlying new thing before adding more intermediate layers.

RossComputerGuy commented:
@Ericson2314 Already done. Afaik, it's built directly into Zig. Maybe I should add an option to disable Zig from using LLVM?

Ericson2314 commented:
@RossComputerGuy I mean package Aro separately and first try making a `stdenvAro`.

RossComputerGuy commented:
I've replaced what I can find using `gcc.arch` and changed a few things around. `lib.systems.architecture` throws an error and I think we can remove it after the next NixOS release. The error message is in place to let people know about the change.

RossComputerGuy commented:
@Ericson2314 Should we do Aro in a different PR or in this one?

Ericson2314 commented:
I'd be fine either way: if it's a separate one, we can just put this one on pause until it's done.

RossComputerGuy commented:
Sounds good, I've decided to include it into this PR to minimize conflicts and just get this PR out the door faster. Should we also package [zld](https://github.com/kubkon/zld) separately (zig ld) in this PR as well?

RossComputerGuy committed:
arocc: init at 0-unstable-06-01

RossComputerGuy commented:
Marked as draft until Aro is implemented properly into nixpkgs.

RossComputerGuy commented:
Arocc is starting to work, need to resolve these first:
```
$ ./result/bin/arocc a.c
warning: unknown warning 'l,-dynamic-linker=/nix/store/rk5wpla6c2idpwidmg8a98lm6n92bfxz-glibc-2.39-52/lib/ld-linux-aarch64.so.1' [-Wunknown-warning-option]
error: unknown argument '-B/nix/store/ir9kmbssp0cjrbryjjxkvgyia7f8kzcp-arocc-wrapper-0-unstable-06-01/bin/'
error: unknown argument '-L/nix/store/rk5wpla6c2idpwidmg8a98lm6n92bfxz-glibc-2.39-52/lib'
```

RossComputerGuy commented:
Aro does somewhat work now but integrated with Nix.

JohnRTitor reviewed (commented):
Please remove the conflicts and (for some reason you have `edits by maintainers disabled`?)

RossComputerGuy commented:
> Please remove the conflicts

Rebased

> (for some reason you have `edits by maintainers disabled`?)

Didn't touch that option when making the PR, it should've been enabled by default like when I made my other PR's. I don't see the option to set it now.

JohnRTitor commented:
Ah, it's because you are using a organisation's cloned Nixpkgs. GitHub only lets maintainers edit the branch if PR author and repo "owner" is the same, presumably.

RossComputerGuy commented:
Seems like something broke upstream with eval?
```
       error: attribute 'dev' missing

       at /ofborg/checkout/2/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/by-name/pi/picocrypt/package.nix:44:5:

           43|     xorg.libXxf86vm
           44|     libGL.dev
             |     ^
           45|     gtk3
```

JohnRTitor reviewed (commented):
Yes, rebase the base branch, should be fixed now.

RossComputerGuy commented:
Rebased

SuperSandro2000 commented on pkgs/by-name/me/meson/package.nix:
```diff
@@ -81,6 +81,13 @@ python3.pkgs.buildPythonApplication rec {
+    # Fix meson's detection for zig's linker
+    (fetchpatch {
+      name = "linker-support-zig-cc.patch";
+      url = "https://github.com/mesonbuild/meson/pull/12293.patch";
```
```suggestion
      url = "https://github.com/mesonbuild/meson/pull/12293/commits/2baae244c995794d9addfe6ed924dfa72f01be82.patch";
```

RossComputerGuy replied:
Done

SuperSandro2000 commented on pkgs/by-name/me/meson/package.nix:
```diff
@@ -81,6 +81,13 @@ python3.pkgs.buildPythonApplication rec {
       hash = "sha256-uSilNuSx9yd1cxs0XVLcLw4MOXEd2uIe2g+wk+SBqeU=";
     })
+
+    # Fix meson's detection for zig's linker
```
```suggestion
    # Fix meson's detection for zig's linker
    # https://github.com/mesonbuild/meson/pull/12293
```

RossComputerGuy replied:
Done

SuperSandro2000 reviewed (commented):
the linked PR is not merged yet

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4278

tomberek commented:
```
x86_64-unknown-linux-gnu-nolibc-gcc> ../../gcc-13.3.0/gcc/analyzer/diagnostic-manager.cc: In member function 'void ana::saved_diagnostic::dump_as_dot_node(pretty_printer*) const':
x86_64-unknown-linux-gnu-nolibc-gcc> ../../gcc-13.3.0/gcc/analyzer/diagnostic-manager.cc:802:28: warning: unknown conversion type character 'E' in format [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wformat=-Wformat=8;;]
x86_64-unknown-linux-gnu-nolibc-gcc>   802 |     pp_printf (pp, "var: %qE\n", m_var);
x86_64-unknown-linux-gnu-nolibc-gcc>       |                            ^
x86_64-unknown-linux-gnu-nolibc-gcc> ../../gcc-13.3.0/gcc/analyzer/diagnostic-manager.cc:802:20: warning: too many arguments for format [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wformat-extra-args-Wformat-extra-args8;;]
x86_64-unknown-linux-gnu-nolibc-gcc>   802 |     pp_printf (pp, "var: %qE\n", m_var);
x86_64-unknown-linux-gnu-nolibc-gcc>       |                    ^~~~~~~~~~~~
error: build of '/nix/store/yr237v031vhn8qwzivg4p2mjh6581nyk-llvm-18.1.8.drv' on 'ssh-ng://nixbld@fd7a:115c:a1e0::22' failed: builder for '/nix/store/yr237v031vhn8qwzivg4p2mjh6581nyk-llvm-18.1.8.drv' failed with exit code 1;
       last 25 log lines:
       > 0x87fcf9 cp_parser_declaration_seq_opt(cp_parser*)
       >         ???:0
       > 0x880196 cp_parser_namespace_definition(cp_parser*)
       >     ???:0
       > 0x880b44 cp_parser_declaration(cp_parser*, tree_node*)
       >  ???:0
       > 0x8812d9 c_parse_file()
       >         ???:0
       > 0x977021 c_common_parse_file()
       >  ???:0
       > Please submit a full bug report, with preprocessed source (by using -freport-bug).
       > Please include the complete backtrace with any bug report.
       > See <https://gcc.gnu.org/bugs/> for instructions.
       > [2359/4385] Building CXX object lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64A57FPLoadBalancing.cpp.o
       > [2360/4385] Building CXX object lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/GISel/AArch64PreLegalizerCombiner.cpp.o
       > [2361/4385] Building CXX object lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/GISel/AArch64PostLegalizerCombiner.cpp.o
```

tomberek commented:
> /nix/store/yr237v031vhn8qwzivg4p2mjh6581nyk-llvm-18.1.8.drv

https://dpaste.com/5MM6CN38C

tomberek commented:
Not sure why yet, but the Zig stdenv seems to perform classical ./configure operations very slowly.

RossComputerGuy commented:
Well @tomberek, I have your answer to that https://github.com/NixOS/nixpkgs/pull/318034/files#diff-145f2d578ddec955b3b15a9c780b174bceb9959f496e404185bb2dd63506af6dR16

We need to set the cache or it just doesn't build. Probably could make it a static path inside the build top.

tomberek commented:
- builds: hello, zlib, nasm, sqlite, brotli, m4, automake
- fails: libpng, libjpeg

RossComputerGuy commented:
Rebased to fix conflicts

tomberek commented:
Still works, same success/fail as before.

alyssais commented on pkgs/by-name/me/meson/package.nix:
```diff
@@ -90,6 +90,14 @@ python3.pkgs.buildPythonApplication rec {
+    # https://github.com/mesonbuild/meson/pull/12293
+    (fetchpatch {
+      name = "linker-support-zig-cc.patch";
+      url = "https://github.com/mesonbuild/meson/pull/12293/commits/2baae244c995794d9addfe6ed924dfa72f01be82.patch";
```
Random commits in unmerged pull requests are not stable URLs.  If you must use a non-upstream patch like this, it has to be vendored to ensure it can still be built in future.

alyssais replied:
This also doesn't look like it's going to be upstream any time soon.  I think this might have been a bit premature…

RossComputerGuy replied:
Yeah, will pull this patch into #331011

RossComputerGuy commented:
Yeah, not surprised a new stdenv has this. Will be fixing this in #331011.

trofi commented:
`pkgsZig` seems to have a similar problem:

```
$ nix build --no-link -f. pkgsZig.ccacheWrapper
error:
       … while calling a functor (an attribute set with a '__functor' attribute)
         at pkgs/top-level/all-packages.nix:17992:19:
        17991|   # should be owned by user root, group nixbld with permissions 0770.
        17992|   ccacheWrapper = makeOverridable ({ extraConfig, cc }:
             |                   ^
        17993|     cc.override {

       … while evaluating a branch condition
         at lib/customisation.nix:148:7:
          147|     in
          148|       if isAttrs result then
             |       ^
          149|         result // {

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: function 'anonymous lambda' called with unexpected argument 'cc'
       at pkgs/development/compilers/zig/cc.nix:1:1:
            1| {
             | ^
            2|   lib,
```
